cvmfs/server/generate_stats_plots.C

```diff
@@ -0,0 +1,578 @@
+    result =
+      result.Define(
+        "chunks_uploaded",
+        [] (Long64_t added, Long64_t duplicated) { return added - duplicated; },
+        {"chunks_added", "chunks_duplicated"});
+
+    if (result.HasColumn("duration")) {
+      result =
+        result.Define("chunks_uploaded_speed",
+                      getSpeed, {"chunks_uploaded", "duration"})
+              .Define("files_added_speed",
+                      getSpeed, {"files_added", "duration"})
+              .Define("sz_bytes_uploaded_speed",
+                      getSpeed, {"sz_bytes_uploaded", "duration"})
+              .Define("sz_bytes_added_speed",
+                      getSpeed, {"sz_bytes_added", "duration"});
+    }
+  }
+
+  return result;
+}
+
+TGraphMap precompute_graphs(ROOT::RDF::RNode* rdf)
+{
+  auto rdf2 = define_custom_cols(rdf);
+  auto result = TGraphMap();
+  for (auto &&col : rdf2.GetColumnNames()) {
+    if (excluded_cols.find(col) != excluded_cols.end()) {
+      continue;
+    }
+
+    result[col] = rdf2.Graph("x", col);
+  }
+  return result;
+}
+
+void generate_stats_plots(std::string stats_db_path, std::string out_path) {
+  TFile f(out_path.c_str(), "RECREATE");
+  // TODO(jpriessn) change "x" to revision
+  rdfs["publish"]["revision"] = new ROOT::RDF::RNode(
+    ROOT::RDF::MakeSqliteDataFrame(stats_db_path, stmt_publish)
+              .Alias("x", "publish_id"));
+  rdfs["publish"]["daily"] = new ROOT::RDF::RNode(
+    ROOT::RDF::MakeSqliteDataFrame(stats_db_path, stmt_publish_daily)
+              .Define("x", getDayCoord, {"day"}));
+  rdfs["publish"]["weekly"] = new ROOT::RDF::RNode(
+    ROOT::RDF::MakeSqliteDataFrame(stats_db_path, stmt_publish_weekly)
+              .Define("x", getWeekCoord, {"date"}));
+  // rdfs["publish"]["monthly"] = new ROOT::RDF::RNode(
+  //   ROOT::RDF::MakeSqliteDataFrame(stats_db_path, stmt_publish_monthly)
+  //             .Define("x", getMonthCoord, {"month"}));
+
+  // TODO(jpriessn) change "x" to revision
+  auto rdf_gc_rev =
+    new ROOT::RDF::RNode(
+    ROOT::RDF::MakeSqliteDataFrame(stats_db_path, stmt_gc)
+              .Alias("x", "gc_id"));
+  bool garbage_collectible;
+  garbage_collectible = *rdf_gc_rev->Count();
+  if (garbage_collectible) {
+    rdfs["gc"]["revision"] = rdf_gc_rev;
+    rdfs["gc"]["daily"] = new ROOT::RDF::RNode(
+      ROOT::RDF::MakeSqliteDataFrame(stats_db_path, stmt_gc_daily)
+                .Define("x", getDayCoord, {"day"}));
+    rdfs["gc"]["weekly"] = new ROOT::RDF::RNode(
+      ROOT::RDF::MakeSqliteDataFrame(stats_db_path, stmt_gc_weekly)
+                .Define("x", getWeekCoord, {"date"}));
+    // rdfs["gc"]["monthly"] = new ROOT::RDF::RNode(
+    //   ROOT::RDF::MakeSqliteDataFrame(stats_db_path, stmt_gc_monthly)
+    //             .Define("x", getMonthCoord, {"month"}));
+  }
+
+  for (auto const& op : rdfs) {
+    for (auto const& period : op.second) {
+      graphs[op.first][period.first] = precompute_graphs(period.second);
```

Comment:
Could this simply use `period` as non-const reference?

Reply:
I'm not sure if I follow. Do you just want to delete the const keyword on line 383?

Reply:
Ah, my misunderstanding, never mind